From 4e9f28caecdd621245ba22c003ed5168e62ab688 Mon Sep 17 00:00:00 2001
From: yvanliang <yvan6liang@outlook.com>
Date: Fri, 9 Oct 2026 15:43:23 +0800
Subject: [PATCH 04/18] GSI Android 17 compatibility:
 0004-Revert-drop-support-for-V-gsi-on-pixel-5-R-base-kern.patch

Change-Id: Ia0f017870c1c4279aecef0e6eae0f9448a3cd77e
---
 ..._android_server_connectivity_ClatCoordinator.cpp | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/service/jni/com_android_server_connectivity_ClatCoordinator.cpp b/service/jni/com_android_server_connectivity_ClatCoordinator.cpp
index 70991eb..437e78f 100644
--- a/service/jni/com_android_server_connectivity_ClatCoordinator.cpp
+++ b/service/jni/com_android_server_connectivity_ClatCoordinator.cpp
@@ -91,6 +91,11 @@ static void verifyPerms(const char * const path,
 
 #undef ALOGF
 
+bool isGsiImage() {
+    // this implementation matches 2 other places in the codebase (same function name too)
+    return !access("/system/system_ext/etc/init/init.gsi.rc", F_OK);
+}
+
 static const char* kClatdDir = "/apex/com.android.tethering/bin/for-system";
 static const char* kClatdBin = "/apex/com.android.tethering/bin/for-system/clatd";
 
@@ -132,6 +137,14 @@ static void verifyClatPerms() {
 
 #undef V2
 
+    // HACK: Some old vendor kernels lack ~5.10 backport of 'bpffs selinux genfscon' support.
+    // This is *NOT* supported, but let's allow, at least for now, U+ GSI to boot on them.
+    // (without this hack pixel5 R vendor + U gsi breaks)
+    if (isGsiImage() && !bpf::isAtLeastKernelVersion(5, 10, 0)) {
+        ALOGE("GSI with *BAD* pre-5.10 kernel lacking bpffs selinux genfscon support.");
+        return;
+    }
+
     if (fatal) abort();
 }
 
